core/src/main/scala/com/databricks/labs/remorph/coverage/estimation/RuleDefinitions.scala

```diff
@@ -0,0 +1,77 @@
+package com.databricks.labs.remorph.coverage.estimation
+
+import upickle.default._
+
+/**
+ * <p>Defines the rules and their related score for conversion complexity estimation.
+ * </p>
+ * <p>
+ *   The rules are defined as a map of rule name to score, and their descriptions are expected to be
+ *   stored somewhere more relevant to the dashboard reporting system (where they can also be subject
+ *   to i18n/l10n).
+ * </p>
+ * <p>
+ *   Rules that are matched by the analyzer will be used to calculate the complexity of the query in terms
+ *   of how much effort it is to convert it to Databricks SQL and not necessarily how complex the query is in
+ *   terms of say execution time or resource requirements. While there are rules to score for the inability
+ *   to parse, generate IR and transpile, they are essentially capturing work for the core team rather than
+ *   the user/porting team. Such scores can optionally be ruled out of conversion complexity calculations but
+ *   are useful to assess the work required from the core Remorph team.
+ * </p>
+ */
+class RuleDefinitions {
+
+  // TODO: Maybe we should use a match like we do in FunctionBuilder?
+
+  val rules: Map[String, Int] = Map(
+    // We were unable to parse the query at all. This adds a significant amount of work to the conversion, but it is
+    // work for the core team, not the user, so are able to filter these out of calculations if desired.
+    "PARSE_FAILURE" -> 100,
```

Comment:
can we instead create the concrete classes, because string constants may get mistyped, causing harder-to-diagnose bugs. e.g.

```
case object ParseFailureRule extends RuleScore(100)
```

also, `case class RuleScore(rule: String, score: Int, from: Seq[RuleScore])` can simply become `Seq[RuleScore]`, right?

Reply:
What we really need here (and in many other places, including most of the foreseeable future developments) is something akin to [this](https://en.wikipedia.org/wiki/Attribute_grammar). 

TLDR; we need the ability to attach arbitrary pieces of information (here it is complexity estimates, but it could as well be some schema information we fetch from an actual database, etc) to the nodes of our precious IR tree, while retaining its structure (because that structure is what we actually care about). 

Once you have this ability, folding such a "labelled-tree" into whatever you need (be it an overall cost estimate, a string containing some SQL code, etc) becomes almost trivial. The best part being that you could label the same tree with different such additional-pieces-of-information at no added cost.

I know how to perform such tree-labelling in a Functional-Progamming-with-recursion-schemes setting, as I applied these techniques with great success on similar problems in the past (namely, transpiling SPARQL queries to Spark-SQL code), though I understand why we won't be considering it as a viable option (it involves writing code that can rapidly become quite difficult to wrap one's head around). However, it would certainly be worth trying to translate these techniques into our more-or-less-imperative setting (even though I cannot guarantee it would be feasible).